### DISCUSSION TITLE
Allow getInitialProps on pages/404

### DISCUSSION THREAD

**Discussion by allan2** [vercel/next.js#17004] [Ideas]
# Feature request

`next-i18next` requires `getInitialProps` to split translation namespaces on Next pages.
The [requirement](https://github.com/vercel/next.js/blob/master/errors/404-get-initial-props.md) to disallow `getInitialProps` on `pages/404` hinders users who wish to have translations on the 404 page itself.

Even if I don't want to use translations on my 404 page, I'd like to be able to specify an empty list of namespaces using `getInitialProps` to satisfy the `next-i18next` HOC requirements.

There is an active discussion over at [https://github.com/isaachinman/next-i18next/issues/677](next-i18next).

## Describe the solution you'd like

Allow `getInitialProps` on `pages/404`. Those who do not want it can choose not to use it.

Perhaps even show a warning about when it is used, with an option to disable it in the config. But this may be overkill.

## Additional context

The `next-i18next` warning for reference only:
```
You have not declared a namespacesRequired array on your page-level component: Custom404.
This will cause all namespaces to be sent down to the client, possibly negatively impacting the performance of your app.
For more info, see: https://github.com/isaachinman/next-i18next#4-declaring-namespace-dependencies
```

**Comment by AliEmili** [comment]
I'm dealing with this issue either. Thanks for informing this as an issue.

**Comment by M1ck0** [comment]
This would be really handy feature to have

Until this is reviewed/approved you can use 

```javascript
module.exports = {
  async redirects() {
    return [
      {
        source: '/_error',
        destination: '/another_error_page',
        permanent: true,
      },
    ]
  },
}
```

And then use `getInitialProps` in that page

This is indeed an extra step that could be avoided by allowing `getInitialProps` for `404` pages but I think that this is the only workaround if you desperately need `i18` for the error page.

You can read more [here](https://nextjs.org/docs/api-reference/next.config.js/redirects)

Not sure if you were aware of this but I'll leave it here because someone might find it helpful :)

**Comment by apt-exploit** [comment]
I am going through the same problem, I join you.

**Comment by KBeDevel** [comment]
Not exactly the solution, but you can get translations working by using `getStaticProps` (credits to [@kachar](https://github.com/i18next/next-i18next/issues/677#issuecomment-832669066)):

> I've stumbled upon this discussion but couldn't find the solution right away here, so I'm posting it for future reference:
> 
> This solution uses `"next-i18next": "^8.2.0"` with `"next": "^10.2.0"`
> 
> ```tsx
> // pages/404.tsx
> import { GetStaticProps } from 'next'
> import NextErrorComponent from 'next/error'
> import { useTranslation } from 'next-i18next'
> import { serverSideTranslations } from 'next-i18next/serverSideTranslations'
> 
> export const getStaticProps: GetStaticProps = async ({ locale }) => {
>   return {
>     props: {
>       ...(await serverSideTranslations(locale ?? 'en', ['common'])),
>     },
>   }
> }
> 
> const NotFoundPage = () => {
>   const { t } = useTranslation()
>   return <NextErrorComponent statusCode={404} title={t('error.title')} />
> }
> 
> export default NotFoundPage
> ```
